common/src/main/java/org/corfudb/common/util/CompatibilityVectorUtils.java

```diff
@@ -0,0 +1,118 @@
+package org.corfudb.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.util.proto.CorfuFeatures.Feature;
+
+import java.util.BitSet;
+
+/**
+ * A collection of {@link BitSet}s representing the Corfu features enabled so far.
+ *
+ * Creates the Compatibility Vectors that contain the information about the Corfu features enabled
+ * so far as per the current stage of the code during the compile time of the class.
+ * The vectors are of type {@link BitSet} in which each bit will represent if the corresponding
+ * feature is enabled or not based on its value (0 or 1).
+ *
+ * Client-Server Transmission:
+ * These BitSet values will be converted to array of bytes (byte[]) set in the Protobuf header
+ * HeaderMsg.ProtocolVersionMsg value (of both RequestMsg and ResponseMsg),
+ * and sent to the receiver. The receiver will then convert the array of bytes into BitSet and check
+ * the specific features using
+ * {@link CompatibilityVectorUtils#isFeatureEnabled(Feature, BitSet)} method.
+ *
+ * Steps for New Features:
+ * 1. <u>Introduce</u> them by creating an entry in the file corfu_features.proto
+ * 2. <u>Enable</u> them by appending to values of the field - 'supportedFeatures' in this class.
+ *
+ * Created by cgudisagar on 5/3/21.
+ */
+@Slf4j
+public class CompatibilityVectorUtils {
+
+    /**
+     * This initializer class loads when getCompatibilityVectors() is called for the first time.
+     * It gives a thread-safe lazy-initialization because the class loader guarantees that all
+     * static initialization is complete before getting access to the class.
+     */
+    @VisibleForTesting
+    private static class CompatibilityVectorWrapper {
+
+        /**
+         * Contains the vectors representing the Corfu features enabled so far.
+         */
+        private static final BitSet vectors = generateVectors();
+
+        private static BitSet generateVectors(){
+            // length-1 to get rid of the default UNRECOGNIZED enum value, else we get an error
+            BitSet bitSet = new BitSet(Feature.values().length-1);
+            Feature[] features = Feature.values();
+            for(int feature_index=1; feature_index < features.length-1; ++feature_index){
+                if(features[feature_index]!= Feature.UNRECOGNIZED ||
```

Comment:
Where is Feature.UNRECOGNIZED defined? I only see one enum that is INVALID in the Feature enum. Am I looking at a wrong commit?

Reply:
> The protocol buffer compiler will generate a Java enum type called Foo with the same set of values. If you are using proto3, it also adds the special value UNRECOGNIZED to the enum type.

[Docs Link](https://developers.google.com/protocol-buffers/docs/reference/java-generated#enum)

Reply:
It seems like UNRECOGNIZED and INVALID are serving the same purpose?

Reply:
We had an offline discussion and changed the enum implementations from Protobuf to Java.